### PR TITLE
Station(Ship) Traits

### PR DESCRIPTION
**Pull request by EgorDinamit**
## About the Pull Request

Partially a port of #tgstation/tgstation/pull/56309
Adds station(ship) traits that will affect the rounds *randomly*.

Doesn't have much traits at the moment, but works regardless.

## Why It's Good For The Game

Makes each new round somewhat different from the previous one, potentially making it easier/harder/funnier/weirder and so on.

### Positive
- Exploration grant: Supply will start with 250 points instead of 50.
- Quick shuttle: Supply drone will take less time to arrive.
- Premium survival boxes: Survival boxes will start with additional equipment. Right now it's painkillers and bruise packs.

### Neutral
- Unnatural atmosphere: Exoplanets will be more likely to have dangerous/unnatural gases.

### Negative
- Slow shuttle: Supply drone will be slower and take more time to arrive.
- Disabled lighting: Some APCs on the ship will start with disabled lighting.

## Did you test it?

I did test it and it works as expected. Would still have to add more traits.

## Changelog

:cl:
rscadd: Each new shift may have different station(ship) traits. Read command report at round-start to see what you got.
/:cl:

### CODE REVIEW

**Comment by Cupax3** [comment]
If this is WIP, make it a draft.
Also, please compile a text about all traits in some form for easier review.

**Comment by Cupax3** [comment]
Has the time testmerged caused any issues?

**Comment by EgorDinamit** [comment]
> Has the time testmerged caused any issues?

The only real issue I got at the moment is that there is only one neutral trait, which may cause runtimes if subsystem attempts to take 2 neutral traits.
Working on a fix rn.

**Comment by EgorDinamit** [comment]
Good to go now.

File: code/controllers/subsystems/processing/station.dm [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	wait = 5 SECONDS
+
+	///A list of currently active station traits
+	var/list/station_traits = list()
+	///Assoc list of trait type || assoc list of traits with weighted value. Used for picking traits from a specific category.
+	var/list/selectable_traits_by_types = list(STATION_TRAIT_POSITIVE = list(), STATION_TRAIT_NEUTRAL = list(), STATION_TRAIT_NEGATIVE = list())
+
+/datum/controller/subsystem/processing/station/Initialize()
+
+	//If doing unit tests we don't do none of that trait shit ya know?
```
Comment (Cupax3) [inline code/controllers/subsystems/processing/station.dm]: Might as well keep the comments clean if we put effort in.
```suggestion
	//If doing unit tests we don't do none of that trait stuff ya know?
```

File: code/controllers/subsystems/processing/station.dm [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+		if(!picked_trait) // Empty!
+			break
+		selectable_traits_by_types[trait_type] -= picked_trait // No duplicates
+		picked_trait = new picked_trait()
+		station_traits += picked_trait
+		if(!picked_trait.blacklist)
+			continue
+		for(var/i in picked_trait.blacklist)
+			var/datum/station_trait/trait_to_remove = i
+			selectable_traits_by_types[initial(trait_to_remove.trait_type)] -= trait_to_remove
```
Comment (Cupax3) [inline code/controllers/subsystems/processing/station.dm:52]: Wouldn't this have a bias toward traits that are indexed earlier?

**Review by Cupax3 — CHANGES_REQUESTED** [review]

File: code/controllers/subsystems/processing/station.dm (hunk 2 above)
Comment (EgorDinamit) [reply code/controllers/subsystems/processing/station.dm:52]: There is no bias, it's simply removing all the traits that were blacklisted by previously selected trait.

File: code/controllers/subsystems/processing/station.dm (hunk 1 above)
Comment (EgorDinamit) [reply code/controllers/subsystems/processing/station.dm]: Didn't want to edit comments because it was a port, but now it's done.

**Comment by EgorDinamit** [comment]
@Cupax3
Updated.

**Review by Cupax3 — APPROVED** [review]

**Pull request merged**